### PR TITLE
Java.Lang.Math

### PR DESCRIPTION
**Pull request by TuanAnhNguyen69**

### CODE REVIEW

File: java/lang/ArithmeticException/ArithmeticException.hpp [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+#ifndef NATIVE_ARITHMETICEXCEPTION_H
+#define NATIVE_ARITHMETICEXCEPTION_H
+
+#include "../String/String.hpp"
+#include "../Exception/Exception.hpp"
+
+namespace Java {
+    namespace Lang {
+        //TODO change to RuntimeException when merge
```
Comment (dthongvl) [inline java/lang/ArithmeticException/ArithmeticException.hpp]: Please change bro

File: java/lang/ArithmeticException/ArithmeticExceptionTest.cpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    // Constructs a new ArithmeticException with the specified detail message.
+    ArithmeticException arithmeticExceptionWithMessage = ArithmeticException("ArithmeticException with the specified message");
+    ASSERT_STR("ArithmeticException with the specified message", arithmeticExceptionWithMessage.getMessage().toString());
+}
+
+TEST (JavaLang, ArithmeticExceptionTryCatch) {
+    try {
+        throw ArithmeticException("Throw ArithmeticException");
+    } catch (Exception e) {
```
Comment (dthongvl) [inline java/lang/ArithmeticException/ArithmeticExceptionTest.cpp]: Please catch ArithmeticException &

File: java/lang/Integer/Integer.hpp [hunk 3]
```diff
@@ -45,6 +45,17 @@ namespace Java {
 			string string_original;
 
 		public:
+			/**
+			 *A constant holding the minimum value an int can have, -2^31.
```
Comment (dthongvl) [inline java/lang/Integer/Integer.hpp]: Add space

File: java/lang/Float/Float.hpp [hunk 4]
```diff
@@ -40,6 +40,89 @@ namespace Java {
...
+             */
+            static constexpr float MIN_VALUE = 0x0.000002P-126f; // 1.4e-45f
+
+            /**
+             * The number of bits used to represent a float value.
+             */
+            static const int SIZE = 32;
+
+            /**
+             *             The number of bytes used to represent a float value.
```
Comment (dthongvl) [inline java/lang/Float/Float.hpp]: Please remove space

File: java/lang/Math/Math.cpp [hunk 5]
```diff
@@ -25,73 +25,464 @@
...
-long Math::abs(long a) {
+double Math::abs(double a) {
 	return a >= 0 ? a : -a;
 }
 
+double Math::acos(double value) {
+    return math_acos(value);
+}
+
+double Math::asin(double a) {
```
Comment (dthongvl) [inline java/lang/Math/Math.cpp]: Please change **a** to **value** or **target**

File: java/lang/Math/Math.cpp [hunk 6]
```diff
@@ -25,73 +25,464 @@
...
+}
+
+float Math::nextAfter(float start, double direction) {
+    return math_nexttowardf(start, direction);
+}
+
+double Math::nextDown(double a) {
+    if (Double::isNaN(a) || a == Double::NEGATIVE_INFINITY) {
+        return a;
+    } else {
```
Comment (dthongvl) [inline java/lang/Math/Math.cpp]: We dont need else here

File: java/lang/Math/Math.cpp [hunk 7]
```diff
@@ -25,73 +25,464 @@
...
+        if (a == 0.0) {
+            return -Double::MIN_VALUE;
+        } else {
```
Comment (dthongvl) [inline java/lang/Math/Math.cpp]: And here

File: java/lang/Math/Math.cpp [hunk 8]
```diff
@@ -25,73 +25,464 @@
...
+            return Double::longBitsToDouble(Double::doubleToRawLongBits(a) + ((a > 0.0) ? -1L : +1L));
+        }
+    }
+}
+
+// TODO need Float.intBitsToFloat, Float.floatToRawIntBits
+float Math::nextDown(float a) {
+    if (Float::isNaN(a) || a == Float::NEGATIVE_INFINITY) {
+        return a;
+    } else {
```
Comment (dthongvl) [inline java/lang/Math/Math.cpp]: Here too

File: java/lang/Math/Math.cpp [hunk 9]
```diff
@@ -25,73 +25,464 @@
...
+float Math::scalb(float a, int scaleFactor) {
+    return math_ldexpf(a, scaleFactor);
+}
+
+double Math::scalb(double a, int scaleFactor) {
+    return math_ldexp(a, scaleFactor);
+}
+
+double Math::signum(double a) {
+    return (a == 0.0 || Double::isNaN(a)) ? a: copySign(1.0, a);
```
Comment (dthongvl) [inline java/lang/Math/Math.cpp]: Please rewrite with full if

File: java/lang/Math/Math.cpp [hunk 10]
```diff
@@ -25,73 +25,464 @@
...
+    }
+    int exp = getExponent(a);
+    switch (exp) {
+        case Double::MAX_EXPONENT + 1:        // NaN or infinity
+            return Math::abs(a);
+        case Double::MIN_EXPONENT - 1:        // zero or subnormal
+            return Double::MIN_VALUE;
+        default:
+            if (exp > Double::MAX_EXPONENT || exp < Double::MIN_EXPONENT) {
+                // TODO Throw Assertion error
```
Comment (dthongvl) [inline java/lang/Math/Math.cpp]: Please throw

**Review by dthongvl — CHANGES_REQUESTED** [review]

File: java/lang/Math/Math.cpp (hunk 10 above)
Comment (TuanAnhNguyen69) [reply java/lang/Math/Math.cpp]: AssertionError is not available

Comment (dthongvl) [reply java/lang/Math/Math.cpp]: Create :v:

File: java/lang/Math/Math.cpp [hunk 11]
```diff
@@ -26,72 +26,492 @@
...
+
+double Math::pow(double base, double exponent) {
+    return math_pow(base, exponent);
+}
+
+double Math::powerOfTwoD(int exponent) {
+    if (exponent < Double::MIN_EXPONENT || exponent > Double::MAX_EXPONENT) {
+        throw AssertionError(exponent);
+    }
+    return Double::longBitsToDouble((((long) exponent + (long) Double::EXP_BIAS)
```
Comment (dthongvl) [inline java/lang/Math/Math.cpp]: are there anyway to shorten this?

File: java/lang/Math/Math.hpp [hunk 12]
```diff
@@ -24,36 +24,870 @@
...
-			static long max(long a, long b);
-			static double min(double a, double b);
-			static float min(float a, float b);
-			static int min(int a, int b);
-			static double pow(double a, double b);
+            /**
+             * Returns the absolute value of an int value
+             *
+             * @param value
+             * @return a if a is positive, else -a
```
Comment (dthongvl) [inline java/lang/Math/Math.hpp]: Please fix this

File: java/lang/Math/Math.hpp [hunk 13]
```diff
@@ -24,36 +24,870 @@
...
+             * @throw ArithmeticException("integer overflow")
+             * @return int
+             */
+			static int decrementExact(int value);
+
+            /**
+             * Returns Euler's number e raised to the power of a double value
+             *
+             * @param exponent
+             * @return the value (e^a),
```
Comment (dthongvl) [inline java/lang/Math/Math.hpp]: Please fix this

File: kernel/math.h [hunk 14]
```diff
@@ -30,9 +30,38 @@
 #include <stdio.h>
 #include <math.h>
 
-double math_pow(double a, double b);
+double math_pow(double base, double exponent);
 double math_floor(double a);
-long math_round(double a);
+double math_round(double a);
```
Comment (dthongvl) [inline kernel/math.h]: Please rename a tooooooo

**Review by dthongvl — CHANGES_REQUESTED** [review]

File: java/lang/Math/Math.hpp [hunk 15]
```diff
@@ -336,20 +337,20 @@ namespace Java {
...
 
             /**
-             * Returns the floor modulus of the int arguments.
+             * Returns the floor modulus of the long arguments.
              *
              * @param dividend
              * @param divisor
              * @throw ArithmeticException if divisor is 0
-             * @return long
+             * @return he floor modulus of the arguments
```
Comment (dthongvl) [inline java/lang/Math/Math.hpp]: the

**Review by dthongvl — CHANGES_REQUESTED** [review]

**Review by dthongvl — APPROVED** [review]

**Pull request merged**